SSA/Projects/InstCombine/Tactic.lean

```diff
@@ -15,46 +15,24 @@ open Ctxt
 /--
 `simp_alive_case_bash` transforms a goal of the form
   `∀ (x₁ : Option (BitVec _)) ... (xₙ : Option (BitVec _)), ...`
-(for some number of variables `n ≤ 5`, following the hack in `simp_peephole`)
 into a goal about just `BitVec`s, by doing a case distinction on each `Option`.
 
 Then, we `simp`lify each goal, following the assumption that the `none` cases
 should generally be trivial, hopefully leaving us with just a single goal:
 the one where each option is `some`. -/
-macro "simp_alive_case_bash" : tactic =>
-  `(tactic|
-    (
-      /- Attempt to introduce up to 5 variables (using `try` because the intro might fail if
-         the goal has less than 5 universal quantifiers)
-
-        case split on them, and simp
-          through the monadic bind on `Option` (in the generally true assumption that the `none`
-            case becomes trivial and is closed by `simp`). -/
-      try intros v0; try intros v1; try intros v2; try intros v3; try intros v4
-
-      /- Then, case split on each variable, and `simp` through the monadic bind on `Option`
-         (in the hope that the `none` case becomes trivial and is immediately closed) -/
-      try cases' v0 with x0 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> try cases' v1 with x1 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> try cases' v2 with x2 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> try cases' v3 with x3 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> try cases' v4 with x4 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> try cases' v5 with x5 <;> simp [Option.bind, bind, Monad.toBind]
-          <;> dsimp [Option.bind, bind, Monad.toBind]
-
-      /- Revert the variables introduced in the `some` cases, so that we are left with
-         a universally quantified goal of the form:
-         `∀ (x₁ : BitVec _) ... (xₙ : BitVec _), ...` -/
-      try revert x5
-      try revert x4
-      try revert x3
-      try revert x2
-      try revert x1
-      try revert x0
-
-      try revert w
-    )
-  )
+syntax "simp_alive_case_bash" : tactic
+macro_rules
+  | `(tactic| simp_alive_case_bash) => `(tactic|
+    try (
+      intro (v : Option (_))    -- Introduce the variable,
+      rcases v with _|x         -- Do the case distinction
+      <;> simp only [BitVec.Refinement.refl, BitVec.Refinement.some_some, bind_assoc]
+      --  ^^^^^^^^^^^^^^^^^^^^^^^^ Simplify, in the hopes that the `none` case is trivially closed
+      <;> simp_alive_case_bash  -- Recurse, to case bash the next variable (if it exists)
+      <;> revert x              -- Finally, revert the variable we got in the `some` case, so that
+                                --   we are left with a universally quantified goal of the form:
+                                --   `∀ (x₁ : BitVec _) ... (xₙ : BitVec _), ...`
+    ))
```

Comment:
```suggestion
macro_rules
  | `(tactic| simp_alive_case_bash) => `(tactic|
    repeat match goal with
    | `(fun ($v : Option (_)) => _) => 
       rcases $v with _|x
       <;> simp only [BitVec.Refinement.refl, BitVec.Refinement.some_some, bind_assoc]
       <;> revert x     
  )
```

Reply:
(Note: Writing this suggestion to show how match goal would help here)

Reply:
I guess the main benefit of match_goal here is that it doesn't swallow errors like `try` does. We could achieve the same by wrapping the tactic body in:
```first
    | fail_if_success (intro (v : Option (_)))
    -- If there is no variable to introduce, the tactic succeeds, but does nothing
    -- This is similar to `try`, except `first ...` does not swallow any errors that occur in
    -- the later tactics
    | intro (v : Option _); ...
`